braindecode/models/usleep.py

```diff
@@ -0,0 +1,272 @@
+                )
+        self.block_postskip = nn.Sequential(
+                    nn.Conv1d(in_channels=(
+                            2 * out_channels if with_skip_connection else out_channels),
+                              out_channels=out_channels,
+                              kernel_size=kernel_size,
+                              padding=padding),  # to preserve dimension (check)
+                    nn.ELU(),
+                    nn.BatchNorm1d(num_features=out_channels),
+                )
+
+    def forward(self, x, residual):
+        x = self.block_preskip(x)
+        if self.with_skip_connection:
+            x, residual = _crop_tensors_to_match(x, residual, axis=-1)  # in case of mismatch
+            x = torch.cat([x, residual], axis=1)  # (B, 2 * C, T)
+        x = self.block_postskip(x)
+        return x
+
+
+class USleep(nn.Module):
+    """Sleep staging architecture from Perslev et al 2021.
+
+    U-Net (autoencoder with skip connections) feature-extractor for sleep
+    staging described in [1]_.
+
+    For the encoder ('down'):
+        -- the temporal dimension shrinks (via maxpooling in the time-domain)
+        -- the spatial dimension expands (via more conv1d filters in the
+           time-domain)
+    For the decoder ('up'):
+        -- the temporal dimension expands (via upsampling in the time-domain)
+        -- the spatial dimension shrinks (via fewer conv1d filters in the
+           time-domain)
+    Both do so at exponential rates.
+
+    Parameters
+    ----------
+    in_chans : int
+        Number of EEG or EOG channels. Set to 2 in [1]_ (1 EEG, 1 EOG).
+    sfreq : float
+        EEG sampling frequency. Set to 128 in [1]_.
+    complexity_factor : float
+        Multiplicative factor for number of channels at each layer of the U-Net.
+        Set to sqrt(2) in [1]_.
+    with_skip_connection : bool
+        If True, use skip connections in decoder blocks.
+    n_classes : int
+        Number of classes. Set to 5.
+    input_size_s : float
+        Size of the input, in seconds. Set to 30.
+    apply_softmax : bool
+        If True, apply softmax on output (e.g. when using nn.NLLLoss). Use
+        False if using nn.CrossEntropyLoss.
+
+    References
+    ----------
+    .. [1] Perslev M, Darkner S, Kempfner L, Nikolic M, Jennum PJ, Igel C.
+           U-Sleep: resilient high-frequency sleep staging. npj Digit. Med. 4, 72 (2021).
+           https://github.com/perslev/U-Time/blob/master/utime/models/usleep.py
+    """
+    def __init__(self,
+                 in_chans=2,
+                 sfreq=100,
+                 complexity_factor=np.sqrt(2),
+                 with_skip_connection=True,
+                 n_classes=5,
+                 input_size_s=30,
+                 apply_softmax=False
+                 ):
+        super().__init__()
+
+        self.in_chans = in_chans
+
+        # Harcoded (otherwise dims can break)
+        depth = 10
```

Comment:
@l-omar-chehab Is this the number of blocks? Isn't it 12 in the paper?

Reply:
It's 12 for Usleep because the frequency is 128Hz. With 100Hz it is only possible to take 10 depth because we can divide tensor of size 1

Reply:
Ah got it! Maybe this should be explained in the docstring in case someone wants to use the architecture with sfreq=128?